### PR TITLE
Update `HealthDisplay` components to use DI to attach data source

### PR DESCRIPTION
**Pull request by peppy**
- [x] Depends on https://github.com/ppy/osu/pull/12670 (to avoid conflicts with caching of `ScoreProcessor`).

### CODE REVIEW

File: osu.Game.Tests/Visual/Gameplay/TestSceneSkinnableHealthDisplay.cs [hunk 1]
```diff
@@ -38,23 +43,18 @@ public void TestHealthDisplayIncrementing()
         {
             AddRepeatStep(@"decrease hp", delegate
             {
-                foreach (var healthDisplay in healthDisplays)
-                    healthDisplay.Current.Value -= 0.08f;
+                healthProcessor.Health.Value = 0.08f;
```
Comment (frenzibyte) [inline osu.Game.Tests/Visual/Gameplay/TestSceneSkinnableHealthDisplay.cs:46]:
```suggestion
                healthProcessor.Health.Value -= 0.08f;
```

File: osu.Game.Tests/Visual/Gameplay/TestSceneSkinnableHealthDisplay.cs [hunk 2]
```diff
@@ -38,23 +43,18 @@ public void TestHealthDisplayIncrementing()
...
             }, 10);
 
             AddRepeatStep(@"increase hp without flash", delegate
             {
-                foreach (var healthDisplay in healthDisplays)
-                    healthDisplay.Current.Value += 0.1f;
+                healthProcessor.Health.Value = 0.1f;
```
Comment (frenzibyte) [inline osu.Game.Tests/Visual/Gameplay/TestSceneSkinnableHealthDisplay.cs:51]:
```suggestion
                healthProcessor.Health.Value += 0.1f;
```

File: osu.Game.Tests/Visual/Gameplay/TestSceneSkinnableHealthDisplay.cs [hunk 3]
```diff
@@ -38,23 +43,18 @@ public void TestHealthDisplayIncrementing()
...
             }, 3);
 
             AddRepeatStep(@"increase hp with flash", delegate
             {
-                foreach (var healthDisplay in healthDisplays)
-                {
-                    healthDisplay.Current.Value += 0.1f;
-                    healthDisplay.Flash(new JudgementResult(null, new OsuJudgement()));
-                }
+                healthProcessor.Health.Value = 0.1f;
```
Comment (frenzibyte) [inline osu.Game.Tests/Visual/Gameplay/TestSceneSkinnableHealthDisplay.cs:56]:
```suggestion
                healthProcessor.Health.Value += 0.1f;
```

File: osu.Game.Tests/Visual/Gameplay/TestSceneSkinnableHealthDisplay.cs [hunk 4]
```diff
@@ -38,23 +43,18 @@ public void TestHealthDisplayIncrementing()
...
+                healthProcessor.ApplyResult(new JudgementResult(new HitCircle(), new OsuJudgement()));
```
Comment (frenzibyte) [inline osu.Game.Tests/Visual/Gameplay/TestSceneSkinnableHealthDisplay.cs:57]: needs to satisfy the conditions in health display (currently doesn't satisfy `judgement.IsHit`):
https://github.com/ppy/osu/blob/755588258e90883f22a89cee219d6e44373b5887/osu.Game/Screens/Play/HUD/HealthDisplay.cs#L42

```suggestion
                healthProcessor.ApplyResult(new JudgementResult(new HitCircle(), new OsuJudgement())
                {
                    Type = HitResult.Perfect,
                });
```

File: osu.Game/Screens/Play/Player.cs [hunk 5]
```diff
@@ -203,9 +203,13 @@ private void load(AudioManager audio, OsuConfigManager config, OsuGameBase game)
             ScoreProcessor.ApplyBeatmap(playableBeatmap);
             ScoreProcessor.Mods.BindTo(Mods);
 
+            dependencies.CacheAs(ScoreProcessor);
+
             HealthProcessor = ruleset.CreateHealthProcessor(playableBeatmap.HitObjects[0].StartTime);
             HealthProcessor.ApplyBeatmap(playableBeatmap);
 
+            dependencies.CacheAs(HealthProcessor);
```
Comment (frenzibyte) [inline osu.Game/Screens/Play/Player.cs:211]: Same as in https://github.com/ppy/osu/pull/12670#issuecomment-835057135, can be exposed as a read-only interface.

Comment (frenzibyte) [reply osu.Game/Screens/Play/Player.cs:211]: Resolving this to be done in a separate PR.

**Review by frenzibyte — DISMISSED** [review]

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**